From 2ea0969e1283d38ed916744d0e223cb9f2fe5846 Mon Sep 17 00:00:00 2001
From: Drew Moseley <drew.moseley@northern.tech>
Date: Sat, 18 Jan 2020 03:11:31 +0000
Subject: [PATCH] mok2verify-multiboot

---
 grub-core/loader/multiboot.c | 29 +++++++++++++++++++++++++++++
 1 file changed, 29 insertions(+)

diff --git a/grub-core/loader/multiboot.c b/grub-core/loader/multiboot.c
index bd9d5b3..bf6fa80 100644
--- a/grub-core/loader/multiboot.c
+++ b/grub-core/loader/multiboot.c
@@ -47,6 +47,7 @@ GRUB_MOD_LICENSE ("GPLv3+");
 
 #ifdef GRUB_MACHINE_EFI
 #include <grub/efi/efi.h>
+#include <grub/efi/mok2verify.h>
 #endif
 
 struct grub_relocator *grub_multiboot_relocator = NULL;
@@ -310,6 +311,20 @@ grub_cmd_multiboot (grub_command_t cmd __attribute__ ((unused)),
   if (! file)
     return grub_errno;
 
+#if GRUB_MACHINE_EFI
+  err = grub_verify_file (argv[0]);
+  if (err != GRUB_ERR_NONE)
+    {
+      grub_error(err, N_("Failed to verify module %s"), argv[0]);
+
+      /* An unauthenticated module always causes a complete boot failure. */
+      if (grub_is_secured () == 1)
+	grub_loader_unset();
+
+      return err;
+    }
+#endif
+
   grub_dl_ref (my_mod);
 
   /* Skip filename.  */
@@ -379,6 +394,20 @@ grub_cmd_module (grub_command_t cmd __attribute__ ((unused)),
   if (! file)
     return grub_errno;
 
+#if GRUB_MACHINE_EFI
+  err = grub_verify_file (argv[0]);
+  if (err != GRUB_ERR_NONE)
+    {
+      grub_error(err, N_("Failed to verify module %s"), argv[0]);
+
+      /* An unauthenticated module always causes a complete boot failure. */
+      if (grub_is_secured () == 1)
+	grub_loader_unset();
+
+      return err;
+    }
+#endif
+
 #ifndef GRUB_USE_MULTIBOOT2
   lowest_addr = 0x100000;
   if (grub_multiboot_quirks & GRUB_MULTIBOOT_QUIRK_MODULES_AFTER_KERNEL)
-- 
2.25.0

